core/src/apps/common/passphrase.py

```diff
@@ -0,0 +1,82 @@
+from micropython import const
+
+import storage.device
+from trezor import wire
+from trezor.messages import ButtonRequestType
+from trezor.messages.ButtonAck import ButtonAck
+from trezor.messages.ButtonRequest import ButtonRequest
+from trezor.messages.PassphraseAck import PassphraseAck
+from trezor.messages.PassphraseRequest import PassphraseRequest
+from trezor.ui import ICON_CONFIG
+from trezor.ui.passphrase import CANCELLED, PassphraseKeyboard
+from trezor.ui.popup import Popup
+from trezor.ui.text import Text
+
+if __debug__:
+    from apps.debug import input_signal
+
+_MAX_PASSPHRASE_LEN = const(50)
+
+
+def is_enabled() -> bool:
+    return storage.device.is_passphrase_enabled()
+
+
+async def get(ctx: wire.Context) -> str:
+    if is_enabled():
+        return await request_from_user(ctx)
+    else:
+        return ""
+
+
+async def request_from_user(ctx: wire.Context) -> str:
+    passphrase = await _get_from_user(ctx)
+    if len(passphrase) > _MAX_PASSPHRASE_LEN:
+        raise wire.DataError("Maximum passphrase length is %d" % _MAX_PASSPHRASE_LEN)
+
+    return passphrase
+
+
+async def _get_from_user(ctx: wire.Context) -> str:
+    if storage.device.get_passphrase_always_on_device():
+        return await request_from_user_on_device(ctx)
+
+    await _entry_dialog()
```

Comment:
starting at this line, until the end, this should be a separate function `_request_from_host` or something like that